File: 2001-3000/2581-count-number-of-possible-root-nodes/2581-count-number-of-possible-root-nodes.cpp
class Solution {
public:
    void dfs(vector<pair<int, bool>> adj[], int i, int p, vector<int> &guessCount) {
        guessCount[i] = 0;
        for (auto &x: adj[i]) {
            if (x.first == p) continue;
            guessCount[i] += x.second;
            dfs(adj, x.first, i, guessCount);
            guessCount[i] += guessCount[x.first];
        }
    }
    void dfs2(vector<pair<int, bool>> adj[], int i, int p, vector<int> &guessCount) {
        for (auto &x: adj[i]) {
            if (x.first == p) continue;
            guessCount[x.first] = guessCount[i] - x.second;
            for (auto &y: adj[x.first]) {
                if (y.first == i && y.second)
                    guessCount[x.first]+=1;
            }
            dfs2(adj, x.first, i, guessCount);
        }
    }
    int rootCount(vector<vector<int>>& edges, vector<vector<int>>& guesses, int k) {
        int n = edges.size()+1;
        vector<pair<int, bool>> adj[n];
        vector<int> guessCount(n);
        for(auto &x: edges) {
            adj[x[0]].push_back({x[1],false});
            adj[x[1]].push_back({x[0],false});
        }

        for (const auto &x: guesses) {
            for (auto &y: adj[x[0]]) {
                if (y.first == x[1])
                    y.second = true;
            }
        }
        dfs(adj, 0, -1, guessCount);
        cout<<guessCount[0]<<endl;
        dfs2(adj, 0, -1, guessCount);
        int ans = 0;
        for (auto &x: guessCount) {
            if (x >= k)
                ans+=1;
        }
        return ans;
    }
};

File: 2001-3000/2581-count-number-of-possible-root-nodes/2581-count-number-of-possible-root-nodes.java
class Solution {
    public int rootCount(int[][] edges, int[][] guesses, int k) {
        int n = edges.length + 1;
        Set<Integer>[] graph1 = new HashSet[n];
        Set<Integer>[] graph2 = new HashSet[n];
        Map<Integer, Integer> parent = new HashMap<>();
        for(int i = 0; i < n; i++) {
            graph1[i] = new HashSet<>();
            graph2[i] = new HashSet<>();
        }
        for(int[] e : edges) {
            int u = e[0], v = e[1];
            if(u > v) {
                int temp = u;
                u = v;
                v = temp;
            }
            graph1[u].add(v);
            parent.put(v, u);
        }
        int[] correctGuesses =  new int[n];
        for(int[] g : guesses) {
            int u = g[0], v = g[1];
            if(graph1[u].contains(v)) {
                correctGuesses[0]++;
            }
            graph2[u].add(v);
        }
        for(int u = 1; u < n; u++) {
            Integer p = parent.get(u);
            if(p == null) {
                continue;
            }
            if((graph2[u].contains(p) && graph2[p].contains(u)) ||
               (!graph2[u].contains(p) && !graph2[p].contains(u))) {
                correctGuesses[u] = correctGuesses[p];
            } else if(graph2[u].contains(p) && !graph2[p].contains(u)) {
                correctGuesses[u] = correctGuesses[p] + 1;
            } else if(!graph2[u].contains(p) && graph2[p].contains(u)) {
                correctGuesses[u] = correctGuesses[p] - 1;
            }
        }
        int res = 0;
        for(int c : correctGuesses) {
            if(c >= k) {
                res++;
            }
        }
        return res;
    }
}